sdk/search/azure-search/src/main/java/com/azure/search/SearchIndexAsyncClient.java

```diff
@@ -0,0 +1,831 @@
+     *
+     * @param documents collection of documents to be merged, if exists, otherwise uploaded
+     * @return response containing the document index result.
+     */
+    public Mono<Response<IndexDocumentsResult>> mergeOrUploadDocumentsWithResponse(Iterable<?> documents) {
+        try {
+            return withContext(context -> mergeOrUploadDocumentsWithResponse(documents, context));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    Mono<Response<IndexDocumentsResult>> mergeOrUploadDocumentsWithResponse(Iterable<?> documents, Context context) {
+        try {
+            IndexBatch<?> batch = buildIndexBatch(documents, IndexActionType.MERGE_OR_UPLOAD);
+            return this.indexWithResponse(batch, context);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    /**
+     * Deletes a collection of documents from the target index
+     *
+     * @param documents collection of documents to delete from the target Index.
+     * @return document index result.
+     */
+    public Mono<IndexDocumentsResult> deleteDocuments(Iterable<?> documents) {
+        try {
+            return this.deleteDocumentsWithResponse(documents)
+                .map(Response::getValue);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    /**
+     * Deletes a collection of documents from the target index
+     *
+     * @param documents collection of documents to delete from the target Index.
+     * @return response containing the document index result.
+     */
+    public Mono<Response<IndexDocumentsResult>> deleteDocumentsWithResponse(Iterable<?> documents) {
+        try {
+            return withContext(context -> deleteDocumentsWithResponse(documents, context));
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    Mono<Response<IndexDocumentsResult>> deleteDocumentsWithResponse(Iterable<?> documents, Context context) {
+        try {
+            IndexBatch<?> batch = buildIndexBatch(documents, IndexActionType.DELETE);
+            return this.indexWithResponse(batch, context);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    /**
+     * Gets Client Api Version.
+     *
+     * @return the apiVersion value.
+     */
+    public SearchServiceVersion getApiVersion() {
+        return this.apiVersion;
+    }
+
+    /**
+     * Gets the endpoint for the Azure Cognitive Search service.
+     *
+     * @return the endpoint value.
+     */
+    public String getEndpoint() {
+        return this.endpoint;
+    }
+
+    /**
+     * Gets the number of documents
+     *
+     * @return the number of documents.
+     */
+    public Mono<Long> getDocumentCount() {
```

Comment:
The document count can never be null. We should change this from `Long` to `long` before shipping a preview. Not blocking this PR.